tests/base.py

```diff
@@ -0,0 +1,25 @@
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License
+# 2.0; you may not use this file except in compliance with the Elastic License
+# 2.0.
+
+"""Shared resources for tests."""
+
+import unittest
+
+from detection_rules import rule_loader
+from detection_rules.rule import Rule
+
+
+class RuleLoadTest(unittest.TestCase):
+    """Base class for shared test cases which need to load rules"""
+
+    @classmethod
+    def setUpClass(cls):
+        cls.rule_files = rule_loader.load_rule_files(verbose=False)
```

Comment:
Does this get called on subclasses too? If so, you can do:
```python
if cls == RuleLoadTest:
  ...
```

Reply:
well in this case we _want_ it loaded in subclasses. Adding that check fails every test that calls any of those attributes. Tested it and it only calls once per TestCase class, regardless of inheritence

Reply:
I think... You want it assigned to each subclass, but you only want to do the loading once. Unless the functions are already cached and the additional loader calls are cheap

Reply:
ah now you are making sense - if that is the case then `setUpClass` may not be what we want at all here, since that is specifically for that. 

```python
_rule_files = rule_loader.load_rule_files(verbose=False)
_rule_lookup = rule_loader.load_rules(verbose=False)
_rules = _rule_lookup.values()
_production_rules = rule_loader.get_production_rules()


class RuleLoadTest(unittest.TestCase):
    """Base class for shared test cases which need to load rules"""

    @classmethod
    def setUpClass(cls):
        cls.rule_files = _rule_files
        cls.rule_lookup = _rule_lookup
        cls.rules = _rules
        cls.production_rules = _production_rules
```

how does this make you feel?

Reply:
alternatively could skip `setUpClass` altogether and just import those straight from `base`

Reply:
if this helps:

# A
`make test  30.07s user 2.49s system 111% cpu 29.129 total`

```python
_rule_files = rule_loader.load_rule_files(verbose=False)
_rule_lookup = rule_loader.load_rules(verbose=False)
_rules = _rule_lookup.values()
_production_rules = rule_loader.get_production_rules()


class RuleLoadTest(unittest.TestCase):
    """Base class for shared test cases which need to load rules"""

    @classmethod
    def setUpClass(cls):
        cls.rule_files = _rule_files
        cls.rule_lookup = _rule_lookup
        cls.rules = _rules
        cls.production_rules = _production_rules
```

# B
`make test  30.56s user 2.62s system 111% cpu 29.848 total`

```python
class RuleLoadTest(unittest.TestCase):
    """Base class for shared test cases which need to load rules"""

    @classmethod
    def setUpClass(cls):
        cls.rule_files = rule_loader.load_rule_files(verbose=False)
        cls.rule_lookup = rule_loader.load_rules(verbose=False)
        cls.rules = cls.rule_lookup.values()
        cls.production_rules = rule_loader.get_production_rules()
```

# C
`make test  30.61s user 2.62s system 110% cpu 30.016 total`

```python
class RuleLoadTest(unittest.TestCase):
    """Base class for shared test cases which need to load rules"""

    rule_files = rule_loader.load_rule_files(verbose=False)
    rule_lookup = rule_loader.load_rules(verbose=False)
    rules = rule_lookup.values()
    production_rules = rule_loader.get_production_rules()
```

Reply:
B > C > A.

Either B or C are good to me.

Reply:
my favorite is this, because you only need to do it once and it will automatically get inherited:
```python
class BaseRuleTest(unittest.TestCase):
    """Base class for shared test cases which need to load rules"""

    @classmethod
    def setUpClass(cls):
        if cls is BaseTestCase:
            cls.rule_files = rule_loader.load_rule_files(verbose=False)
            cls.rule_lookup = rule_loader.load_rules(verbose=False)
            cls.rules = cls.rule_lookup.values()
            cls.production_rules = rule_loader.get_production_rules()
```

Reply:
unittest didn't like this still. Just stuck with B, thanks 👍